File: services/typescript/koa/src/routes/handlers/delete.ts
import type { RouterContext } from "@koa/router";
import { type State, parseId } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";

export function deleteHandler(state: State) {
  return (ctx: RouterContext) => {
    const logger = getLogger();

    const id = parseId(ctx.params.id);
    if (!id.success) {
      logger.warn({ method: ctx.req.method, path: ctx.path }, "Invalid path");
      return sendError(ctx, AppErrors.InvalidPath);
    }

    const task = state.tasks.get(id.data);

    if (task === undefined) {
      logger.warn(
        { id: id.data, method: "DELETE", path: ctx.path },
        "Task not found"
      );
      return sendError(ctx, AppErrors.TaskNotFound);
    }

    state.tasks.delete(id.data);

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        operation: task.operation.toString().toLowerCase(),
        method: "DELETE"
      },
      "Removed task"
    );

    ctx.status = 204;
    ctx.body = null;
  };
}

File: services/python/sanic/src/routes/handlers/delete.py
import structlog
from uuid import UUID
from sanic import Request, empty
from sanic.response import JSONResponse, HTTPResponse

from app.state import AppState
from app.task import Task
from routes.errors import send_error, AppErrors


async def delete_handler(request: Request, id: UUID) -> HTTPResponse | JSONResponse:
    logger = structlog.get_logger()

    state: AppState = request.app.ctx.state

    async with request.app.ctx.tasks_lock:
        task: Task | None = state.tasks.pop(id, None)

        if task is None:
            logger.warn(
                "Task not found", id=str(id), method=request.method, path=request.path
            )
            return send_error(AppErrors.TaskNotFound)

    logger.info(
        "Removed task",
        id=str(id),
        secret=len(task.secret),
        operation=task.operation.lower(),
        method=request.method,
    )

    return empty()
